source/slang/slang-syntax.cpp

```diff
@@ -71,26 +71,117 @@ ABSTRACT_SYNTAX_CLASS(GlobalGenericParamSubstitution, Substitutions);
 #include "slang-val-defs.h"
 #include "slang-object-meta-end.h"
 
-bool SyntaxClassBase::isSubClassOfImpl(SyntaxClassBase const& super) const
+
+/* static*/ SyntaxClassBase::ClassInfo* SyntaxClassBase::ClassInfo::s_first = nullptr;
+
+SyntaxClassBase::ClassInfo::ClassInfo(const char* name, CreateFunc createFunc, const ClassInfo* superClass):
+    m_name(name),
+    m_createFunc(createFunc),
+    m_superClass(superClass),
+    m_next(s_first)
+{
+    m_classId = 0;
+    m_childrenEndClassId = 0;
+
+    s_first = this;
+}
+
+
+
+static uint32_t _calcRangeRec(const SyntaxClassBase::ClassInfo* classInfo, const Dictionary<const SyntaxClassBase::ClassInfo*, List<const SyntaxClassBase::ClassInfo*> >& childMap, uint32_t index)
+{
+    classInfo->m_classId = index++;
+    // Do the calc range for all the children
+    auto list = childMap.TryGetValue(classInfo);
+
+    if (list)
+    {
+        for (auto child : *list)
+        {
+            index = _calcRangeRec(child, childMap, index);
+        }
+    }
+
+    classInfo->m_childrenEndClassId = index;
+    return index;
+}
+
+bool SyntaxClassBase::ClassInfo::isSubClassOfSlow(const ThisType& super) const
 {
-    SyntaxClassBase::ClassInfo const* info = classInfo;
+    SyntaxClassBase::ClassInfo const* info = this;
     while (info)
     {
-        if (info == super.classInfo)
+        if (info == &super)
             return true;
+        info = info->m_superClass;
+    }
+    return false;
+}
+
+static bool _checkSubClassRange()
+{
+    typedef SyntaxClassBase::ClassInfo ClassInfo;
 
-        info = info->baseClass;
+    List<const ClassInfo*> list;
+    for (const ClassInfo* type = ClassInfo::s_first; type; type = type->m_next)
+    {
+        list.add(type);
     }
 
-    return false;
+    for (Index i = 0; i < list.getCount(); ++i)
+    {
+        for (Index j = 0; j < list.getCount(); ++j)
+        {
+            auto a = list[i];
+            auto b = list[j];
+            if (a->isSubClassOf(*b) != a->isSubClassOfSlow(*b))
+            {
+                return false;
+            }
+        }
+    }
+
+    return true;
 }
 
-NodeBase* _dynamicCastImpl(NodeBase* node, SyntaxClassBase const& toClass)
+
+/* static */SlangResult SyntaxClassBase::ClassInfo::initRanges()
 {
-    if(!node) return nullptr;
-    if(node->getClass().isSubClassOfImpl(toClass))
-        return node;
-    return nullptr;
+    // Remove the warning about not referenced
+    SLANG_UNUSED(&_checkSubClassRange);
+
+    // TODO(JS):
+    // Note that the calculating of the ranges could be done more efficiently by adding to an array of struct { super, class }, sorting, by super classs
+    // and using a dictionary to map from class it's first in list of super class use. This works for now though.
+  
+    // We want to produce a map from a node that holds all of it's children
+    Dictionary<const ThisType*, List<const ThisType*> > childMap;
```

Comment:
What seems like an even simpler alternative to me would be that instead of a single global linked list of all the `ClassInfo`s, you could have each `ClassInfo` hold a pointer for a singly-linked list of direct subclasses. Then the constructor for `ClassInfo` would add it to the list of the super class (if any), and you'd get a single tree structure that you could recursively walk in one go to assign the numeric ranges.

Reply:
It's an interesting idea. One problem is construction order - well it's the order they are defined within TU, but here the order the variables are specified does not honor super/child relationship, so I'd be modifying an un-constructed variable. Might be ok - but probly a bit dodgy. But the observation made me make sure that s_first is constructed first. I'd note in practice that the that style works across TU because assigning 0 to a global becomes BSS, and so is 'executed' before anything else - but that specific behavior is not strictly part of C++ standard.

Reply:
Construction order is a good point. Right now all these variables get defined in a single TU, and as far as I know the declaration order is consistent with the inheritance graph (for the simple reason that the same macro magic generated the `class` declarations too, and those need super-classes to come before subclasses also). It probably isn't worth refactoring this to take advantage of any such assumptions since the long-term goal would be to not require this logic in the first place.